### PR TITLE
Improve generated debug info to avoid assertion in SROA

### PR DESCRIPTION
**Pull request by Keno**
This is a very simple start on emitting more thorough debug info. I was planning (and still am) to work on this as part of a larger push to improve our generated debug info, but at the moment SROA is crashing on 3.7 (#13754), because our debug info is not quite valid. Interestingly, I can still see a crash if I run the system image that results from this through bugpoint, the optimizer still crashes, so we might be hitting another LLVM bug here. Bootstrap goes through fine with this though.

### CODE REVIEW

**Comment by Keno** [comment]
Turns out it just runs through the memory cap trying to optimize the system image ;).

**Comment by yuyichao** [comment]
So the "LLVM bug" is that they didn't buy you more memory ;-p

Thanks for the fix. I can confirm that this indeed fixes the bootstrap failure on linux too.

**Comment by Keno** [comment]
Not quite. The problem is that bugpoint limits executions of the optimizer to 400MB, the machine has plenty of RAM.

**Comment by yuyichao** [comment]
> The problem is that bugpoint limits executions of the optimizer to 400MB

I remember you said ["LLVM does not keep such counts"](https://github.com/JuliaLang/julia/issues/11553#issuecomment-111298403). Is this a recent improvement (that we could use for benchmarking) or is it just running it in a separate process and let the OS take care of it (e.g. with `ulimit`).

**Comment by Keno** [comment]
Yes, it's using `ulimit -v` on an optimizer subprocess.

**Comment by Keno** [comment]
Hmm, looks like we have a non-deterministic test failure here :(. Will see if I can reproduce.

File: src/codegen.cpp [hunk 1]
```diff
@@ -4569,7 +4569,7 @@ static Function *emit_function(jl_lambda_info_t *lam)
                 s->name,                // Variable name
                 topfile,                 // File
                 toplineno == -1 ? 0 : toplineno, // Line (for now, use lineno of the function)
-                julia_type_to_di(varinfo.value.typ, ctx.dbuilder, specsig), // Variable type
+                julia_type_to_di(varinfo.value.typ, ctx.dbuilder, varinfo.value.isboxed), // Variable type
```
Comment (vtjnash) [inline src/codegen.cpp]: isboxed isn't really well defined for varinfo.value, but regardless, it hasn't been assigned yet here. i'm not certain that this is what you want here anyways. I think you want to check after emitting the function whether a memloc needed to be created for this variable, or whether it was allocated in `.value` and whether `varinfo.value.typ` is a known type or not.

Comment (Keno) [inline src/codegen.cpp]: I'm considering always passing false here and handing this at the callsite.

**Comment by Keno** [comment]
Ok, now passing `false` everywhere to always describe the actual type rather than the box. Note that I actually found a bug in LLVM where it would mis-lower our debug info (http://reviews.llvm.org/D14186), which however I don't think should cause any crashes, just wrong debug info.

**Comment by tkelman** [comment]
Does this fix #13754? Is it okay to merge?

**Comment by yuyichao** [comment]
This does fix #13754 for me.

**Pull request merged**